Add Upsell section for Premium Lesson Patterns

fjorgemota opened the pull request:
Fixes #5135

### Changes proposed in this Pull Request

* Adapt PatternsStep component to accept injection components between the title and the pattern list (6c9bf40dcbfb9507c35f6c5d58aaa6ffe9b29546);
* Add Upsell for the Premium Lesson Patterns (24a3c554f2cea481343b062d0d5c888136d06f00 and b200fb80b4071da9f6020741b426fa28a361c5ac)

### Testing instructions

On a WordPress installation with the code from this branch:

1. Create a new Lesson;
2. On the Editor Wizard, click "Continue" on the first step;
3. On the Lesson Patterns step, verify that the section "Want more lesson types, check out Sensei Pro." only appears if Sensei Pro is not installed;
4. Verify that the section "Want more lesson types, check out Sensei Pro." match the design (as per this post p6rkRX-3zM-p2);  
5. Verify that clicking the "Learn More" link opens the SenseiLMS.com pricing page in a new tab, and with some UTM tracking code as tracking parameters;

### Screenshot / Video

![Screenshot of the Lesson Patterns step with the Upsell section](https://user-images.githubusercontent.com/529864/172959303-13037c01-6911-42ca-905b-167f77e0b640.png)

fjorgemota committed:
Prepare PatternsStep component to accept a Upsell Section component

fjorgemota committed:
Add Upsell Section to the Lesson Patterns step

fjorgemota committed:
Add utm tracking code and rel attribute to Learn More link

fjorgemota committed:
Remove unneeded space after the title of the section

fjorgemota committed:
Whoops, add missing SVG containing Sensei Tree logo

aaronfc commented on assets/admin/editor-wizard/steps/lesson-patterns-step.js:
```diff
@@ -15,21 +19,67 @@ import PatternsStep from './patterns-step';
-		/>
+		<Fragment>
+			<PatternsStep
+				title={ __( 'Lesson Patterns', 'sensei-lms' ) }
```
Nice catch!

fjorgemota replied:
Hey @aaronfc! Thanks for the review! 

Unfortunately, the title shouldn't be changed because the Figma files aren't updated according to what was decided on Slack. More context on the link p1653529999028269/1653517752.794919-slack-C013QUH20TS, and the fix is on the commit [34fdc93](https://github.com/Automattic/sensei/pull/5254/commits/34fdc93f0025cadabdc16d656aca3f2c29dc3acf)

aaronfc commented on assets/admin/editor-wizard/steps/lesson-patterns-step.js:
```diff
@@ -15,21 +19,67 @@ import PatternsStep from './patterns-step';
+		<Fragment>
+			<PatternsStep
+				title={ __( 'Lesson Patterns', 'sensei-lms' ) }
+				replaces={ replaces }
```
Wouldn't it be easier to pass here a "showBanner" true/false attribute instead of using the Fills/Slots?

fjorgemota replied:
Hmmm, originally, I was passing an argument with a function (or undefined) that, if defined, returned a component to inject between the title and the patterns list.

Adding a boolean option to showBanner could be an alternative, but I feel like the Upsell section should be, for now, dedicated to the Lesson Patterns. Given that the Patterns Step is generic (and shared between the Lesson Patterns and the Course Patterns), such an option could create some confusion while reading the code (e.g. as something that is only shown on the lesson patterns, is saved directly in the generic patterns step component). So I'd be a little against that idea.

Considering that, and that passing an argument with a function just to return a component is also, IMHO, a little weird, I decided to go with slot/fill, which, as far as I understand, [has this exact use-case](https://developer.wordpress.org/block-editor/reference-guides/components/slot-fill/).

WDYT? Do you prefer any of these options more strongly? I'm trying to find the more react-y way of doing this here. :P

aaronfc replied:
Ok, I forgot that the PatternStep is shared between Lessons and Courses.
I'm ok with your approach then :)

aaronfc reviewed (commented):
Looks good. Left a comment about the Slots/Fills usage and also noticed some visual differences:
- First line looks like "stronger" (bold) in the designs. In fact it looks like Sensei Pro is even stronger, but it might be a visual effect.
- Learn more link is blue when hovering. Not sure if that's the expected result but with the black background looks a little bit weird.

fjorgemota committed:
Fix title of the lesson to what was originally discussed

More context: p1653529999028269/1653517752.794919-slack-C013QUH20TS

fjorgemota committed:
Don't change the hover color of the link to blue

Let it be white and remove underline of the link (so the user can see some effect on the link while hovering)

fjorgemota committed:
Make title of the Upsell Section a little stronger

fjorgemota commented:
Hey @aaronfc! Thanks for the review.

I made the title a little stronger, as presented in Figma, by changing the Font Weight. See [6d5e05f](https://github.com/Automattic/sensei/pull/5254/commits/6d5e05f3052550446b3ffb4e01da36f2caf44601).

I also fixed the hover of the link by just removing the underline but maintaining the color the same. See [73a4763](https://github.com/Automattic/sensei/pull/5254/commits/73a47630c57a4db9110c8a4f152b3c68b9728253).

LMK what you think.

aaronfc reviewed (approved):
Looks good and works well! 🚀

alexsanford commented:
Could we tweak the "Learn More" link to open in a new tab?

Also, I find it weird that this checks for Sensei Pro being _installed_ rather than _activated_. I feel like I made this comment on a similar PR previously as well. It's not a big deal (we don't need to make a last-minute change) but I feel like this UI should be based on whether Sensei Pro is actually activated. (I'm talking about plugin activation, not necessarily license activation). Otherwise, if it's installed but not activated, we don't see the upsell prompt, but we also don't get the premium patterns.

Just an observation though. Again, probably not worth a change at this point.

burtrw commented:
For the copy, sorry this didn't get added earlier in the process but, lets roll with:

**Want More Lesson Types?**
Get flashcards, timed quizzes, image hotspots, and more with Sensei Pro. Learn more.

fjorgemota commented:
> Could we tweak the "Learn More" link to open in a new tab?

Actually, yeah. I swear I had added a `target="_blank"` on the Learn More link, but looking again it's...not there? I have no idea what happened, but you can see on the Testing Instructions that I was really expecting it for some reason:

> Verify that clicking the "Learn More" link opens the SenseiLMS.com pricing page **in a new tab**, and with some UTM tracking code as tracking parameters;

I'll open a new PR to fix that.

> Also, I find it weird that this checks for Sensei Pro being installed rather than activated. I feel like I made this comment on a similar PR previously as well. It's not a big deal (we don't need to make a last-minute change) but I feel like this UI should be based on whether Sensei Pro is actually activated. (I'm talking about plugin activation, not necessarily license activation). Otherwise, if it's installed but not activated, we don't see the upsell prompt, but we also don't get the premium patterns.

About this, I think I can change that, but it'll need a change also in the backend, as that information isn't currently included with the extensions data.

I'll take a look. 

> For the copy, sorry this didn't get added earlier in the process but, lets roll with:

No problem, I'll update that in the new PR too. =)
